packages/contracts-bedrock/src/L2/CrossL2Inbox.sol

```diff
@@ -0,0 +1,167 @@
+///         chain. It is permissionless to execute a cross chain message on behalf of any user.
+contract CrossL2Inbox is ICrossL2Inbox, ISemver {
+    /// @notice Selector for 'NotEntered()' error message.
+    ///         Equal to bytes32(uint256(keccak256("NotEntered()")) - 1)
+    uint32 public constant ERR_NOT_ENTERED = 0xbca35af6;
+
+    /// @notice Transient storage slot that `entered` is stored at.
+    ///         Equal to bytes32(uint256(keccak256("crossl2inbox.entered")) - 1)
+    bytes32 public constant ENTERED_SLOT = 0x6705f1f7a14e02595ec471f99cf251f123c2b0258ceb26554fcae9056c389a51;
+
+    /// @notice Transient storage slot that the origin for an Identifier is stored at.
+    ///         Equal to bytes32(uint256(keccak256("crossl2inbox.identifier.origin")) - 1)
+    bytes32 public constant ORIGIN_SLOT = 0xd2b7c5071ec59eb3ff0017d703a8ea513a7d0da4779b0dbefe845808c300c815;
+
+    /// @notice Transient storage slot that the blocknumber for an Identifier is stored at.
+    ///         Equal to bytes32(uint256(keccak256("crossl2inbox.identifier.blocknumber")) - 1)
+    bytes32 public constant BLOCKNUMBER_SLOT = 0x5a1da0738b7fdc60047c07bb519beb02aa32a8619de57e6258da1f1c2e020ccc;
+
+    /// @notice Transient storage slot that the logIndex for an Identifier is stored at.
+    ///         Equal to bytes32(uint256(keccak256("crossl2inbox.identifier.logindex")) - 1)
+    bytes32 public constant LOG_INDEX_SLOT = 0xab8acc221aecea88a685fabca5b88bf3823b05f335b7b9f721ca7fe3ffb2c30d;
+
+    /// @notice Transient storage slot that the timestamp for an Identifier is stored at.
+    ///         Equal to bytes32(uint256(keccak256("crossl2inbox.identifier.timestamp")) - 1)
+    bytes32 public constant TIMESTAMP_SLOT = 0x2e148a404a50bb94820b576997fd6450117132387be615e460fa8c5e11777e02;
+
+    /// @notice Transient storage slot that the chainId for an Identifier is stored at.
+    ///         Equal to bytes32(uint256(keccak256("crossl2inbox.identifier.chainid")) - 1)
+    bytes32 public constant CHAINID_SLOT = 0x6e0446e8b5098b8c8193f964f1b567ec3a2bdaeba33d36acb85c1f1d3f92d313;
+
+    /// @custom:semver 1.0.0
+    string public constant version = "1.0.0";
+
+    /// @notice Returns the origin address of the Identifier. If not entered, reverts.
+    /// @return _origin The origin address of the Identifier.
+    function origin() external view returns (address _origin) {
+        assembly {
+            _entered := tload(ENTERED_SLOT)
+            if eq(_entered, 0) {
+                mstore(0x00, ERR_NOT_ENTERED)
+                revert(0x1C, 0x04)
+            }
+            _origin := tload(ORIGIN_SLOT)
+        }
+    }
+
+    /// @notice Returns the block number of the Identifier. If not entered, reverts.
+    /// @return _blocknumber The block number of the Identifier.
+    function blocknumber() external view returns (uint256 _blocknumber) {
+        assembly {
```

Comment:
You should turn this logic into a modifier or internal function

Reply:
good idea!